Add statistics to output

swashcap opened the pull request:
We’ll eventually merge the _addstatisticoutput_ branch to enable better result reporting for the computation. This PR gets the conversation started.

JingMing623 committed:
add goodness of fit and P value for regression

JingMing623 committed:
add X_label and y_label to return list

swashcap commented on package.json:
```diff
@@ -56,6 +56,7 @@
     "numeric": "^1.2.6",
-    "pify": "^2.3.0"
+    "pify": "^2.3.0",
+    "distributions":"^1.0.0"
```
Going to re-add the _distributions_ package via the CLI. This will modify _package.json_.

swashcap commented on src/index.js:
```diff
@@ -45,13 +48,213 @@ module.exports = {
+       console.log('opts.remoteResult.data.endOfIteration is' , opts.remoteResult.data.endOfIteration);
+       console.log('opts.remoteResult.data.statisticStep is', opts.remoteResult.data.statisticStep);
+        
+        if ((opts.remoteResult.data.endOfIteration == true) && (opts.remoteResult.data.statisticStep == 0)){
```
I’m a little iffy on these branches based on remote result state. Can we separate these into different local computations and let COINSTAC’s pipeline runner do the work?

JingMing623 replied:
Yes, we could merge this part into the local.js, or we can have local_comp1.js, local_comp2.js....... and on the remote node, we can also have remote_comp1.js, local_comp2.js..... ,  maybe you can help me on this. I did plan to write separate code for each condition.......Probably show me an example of modularizing code, I think single shot also can be modularized

swashcap commented on src/index.js:
```diff
@@ -45,13 +48,213 @@ module.exports = {
+           const betaVector = regression.oneShot(biasedX, y);
+           const rSquared = regression.rSquared(biasedX, y, betaVector);
+           const tValue = regression.tValue(biasedX, y, betaVector);
+           var tdist = distributions.Studentt(localCount-1);
```
`var`? Prefer `const` or `let`.

swashcap commented on src/index.js:
```diff
@@ -45,13 +48,213 @@ module.exports = {
+             pValue,
+             biasedX,
+             y,
+             endOfIteration : true,
```
Our style guide prefers no space before the `:` char:

```diff
-             endOfIteration : true,
+             endOfIteration: true,
```

swashcap commented on src/index.js:
```diff
@@ -45,13 +48,213 @@ module.exports = {
+             statisticStep : 0,
+          };          
+        }
+        else if ((opts.remoteResult.data.endOfIteration == true) && (opts.remoteResult.data.statisticStep == 1)){   
```
The fact that we're branching on `statisticStep` makes me want to separate these into local computations, as mentioned before. We can at least move functionality into _src/local.js_ to clean up our computation file.

JingMing623 replied:
Yes, we should separate the computation.

swashcap commented on src/index.js:
```diff
@@ -45,13 +48,213 @@ module.exports = {
+         var totalY=0;
+         var globalYCount=0;
+
+         for (let i=0; i < siteCount; i += 1) {
```
An `Array#reduce` could clean this up a bit.

swashcap commented on src/index.js:
```diff
@@ -45,13 +48,213 @@ module.exports = {
+
+        // get passed parameters from local nodes
+         const currW = opts.previousData.currW;
+         const betaVectorLocal = userResults.map(r => r.data.betaVector);
```
There’s a lot of repeated iteration over `userResults`, [similar to _decentralized-single-shot-ridge-regression_](https://github.com/MRN-Code/decentralized-single-shot-ridge-regression/pull/11#discussion_r96543959).

swashcap committed:
Fix linting errors.

swashcap commented:
I rebased this branch so it has upstream changes. You might need to do a hard reset when merging this branch with your local copy.

swashcap committed:
Move computation functions into respective modules.

swashcap committed:
Clean up local getStatistics function.

JingMing623 commented:
Okay. Will work on resetting tomorrow. Thanks Cory.


> On Jan 19, 2017, at 16:03, Cory Reed <notifications@github.com> wrote:
> 
> I rebased this branch so it has upstream changes. You might need to do a hard reset when merging this branch with your local copy.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

swashcap committed:
Separate computation functionality into functions.

This isolates local and remote computation steps into their own
functions to make code more readable. It also moves p-value generation
into src/regression.js to make code DRYer.

swashcap committed:
Change `console` to `debug`.

swashcap commented:
b51b964217d10bfff47f95e19fc3f35ee2ce6692 changes the computation step separation into separate functions, unified into `remote.run` and `local.run` functions. Test coverage is low, but let's ensure it actually works first.

swashcap commented:
One thing to note: I swapped all `console.log` calls with [debug](https://www.npmjs.com/package/debug) to make runs less chatty. To enable logging you’ll need to ensure the `DEBUG` env var is set:

```shell
$ export DEBUG=coinstac:multishot
$ npm run simulate
```

swashcap committed:
Ensure final remote object is fully output.

swashcap commented:
This PR is pretty outdated. Closing, we can open a new one for refactoring if needed.
